File: Algorithms/229. Majority Element II.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        count1 = 0
        count2 = 0
        can1 = -1234567
        can2 = -1234567
        for i in nums:
            if (i == can1) or (i == can2):
                if(i == can1):
                    count1 += 1
                else:
                    count2 += 1
            else:
                if(count1 == 0):
                    can1 = i
                    count1 = 1
                elif count2 == 0:
                    can2 = i
                    count2 = 1
                else:
                    count1 -= 1
                    count2 -= 1

        count1 = count2 = 0
        for i in nums:
            if i == can1:
                count1 += 1
            if i == can2:
                count2 += 1
        res = []
        if count1 > len(nums)//3:
            res.append(can1)
        if count2 > len(nums)//3:
            res.append(can2)
        return res

File: Algorithms/229. Majority Element II.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int c1=0,c2=0;
        Integer n1=null,n2=null;
        for(int i=0;i<nums.length;i++){
            if((n1==null)||(n1==nums[i])){
                n1=nums[i];
                c1+=1;
            }
            else if ((n2==null)||(n2==nums[i])){
                c2+=1;
                n2=nums[i];
            }
            else if (c1==0){
                n1=nums[i];
                c1+=1;
            }
            else if (c2==0){
                n2=nums[i];
                c2+=1;
            }
            else{
                c1-=1;
                c2-=1;
            }
            // System.out.println(Integer.toString(c1)+c2);
        }
        c1=0;c2=0;
        for(int i=0;i<nums.length;i++){
            if((n1!=null)&&(n1==nums[i])) c1+=1;
            if((n2!=null)&&(n2==nums[i])) c2+=1;
        }
        List<Integer> ans=new ArrayList<Integer>();
        if(c1>nums.length/3) ans.add(n1);
        if((c2>nums.length/3)) ans.add(n2);
        return ans;
    }
}